builder/common/step_create_ssm_tunnel.go

```diff
@@ -75,15 +81,51 @@ func (s *StepCreateSSMTunnel) Run(ctx context.Context, state multistep.StateBag)
 			LocalPort:  s.LocalPortNumber,
 			Region:     s.Region,
 		}.Start(ssmCtx, ui)
-
 		if err != nil {
 			ui.Error(fmt.Sprintf("ssm error: %s", err))
 		}
 	}()
 
+	if len(s.Comm.SSHPrivateKey) > 0 {
+		ui.Say("Uploading public key to instance")
+		err := s.sendUserSSHPublicKey(instance, s.Comm.SSHPrivateKey)
+		if err != nil {
+			ui.Error(err.Error())
+			return multistep.ActionHalt
+		}
+		ui.Say(fmt.Sprintf("Uploaded public key to instance"))
+	}
+
 	return multistep.ActionContinue
 }
 
+func (s *StepCreateSSMTunnel) sendUserSSHPublicKey(instance *ec2.Instance, privateKey []byte) error {
+	publicKeyBytes, err := sshkey.PublicKeyFromPrivate(privateKey)
```

Comment:
I think the SSHPublic key is never set. I'll try it out.

Reply:
[This](https://github.com/hashicorp/packer-plugin-sdk/blob/main/communicator/step_ssh_keygen.go#L43) is the part that set the value for SSHPublicKey. But somehow the value is not there when I try to access it.

Could you check [this commit](https://github.com/hashicorp/packer-plugin-amazon/pull/296/commits/e5e6ce415c4e5efd72bb6acec50093c9dda09574) of how I'm doing this, maybe I'm making a mistake?

Reply:
Hm, looks like the PublicKey is only set when we generate an SSH key, my bad here! Though tbf maybe both should be set when we read a private key from the outside. I'll take a look at the SDK, this could be fixed here for a later release.

In the meantime, I imagine your initial implementation would work, sorry for misleading you on there.

Also, come to think of it, it feels like this should also work when the key is read from a file (see the `SSHPrivateKeyFile` attribute from the `SSHConfig` struct). Unless I'm mistaken, this was not supported in your original implementation, correct?

There's also a `SSHCertificateFile` that may be relevant there too, what do you think?

Considering the multiple possibilities here, the whole `Communicator.SSH` can be passed along, and we can then choose which to upload (or maybe we can upload all the relevant public keys, if that's a possibility?).

Reply:
Yeah I was originally testing with providing the key via file. since I think this issue only happens when you specify the file, otherwise packer/amazon will take care of adding the key to instance.

But I don't this hurts to use the SSHConfig and upload all the public keys that are used in the session to the instance. Also this issue might happen for other cases too, and I was only fixing it for the problem I knew.

The `SSHCertificateFile` seems relevant. I did not know Go had these constructs for working with SSH. I we add an `AuthMethod` from the certs and probably I can scan all the public key auth methods and add them to the instance. This way it works in a more generic way.

Thanks for the suggestion. Will apply & test it tomorrow.

Reply:
So I'm going with your suggestion, passing the `Communicator.SSH` and uploading the public keys I mentioned above.

Reply:
@lbajolet-hashicorp I tried to only pass the `Communicator.SSH` but there's a method on the `Communicator.Config` which is useful here:
```
CommunicatorConfig.ReadSSHPrivateKeyFile()
```

I'm reading the private key file with this method.

Reply:
Since we decided to remove the SSHPrivateKeyFile I'm falling back to only passing the `Communicator.SSH`.